File: ios/Plugin/SQLiteVanillaPlugin.swift
/**
 * (c) 2026, Micro:bit Educational Foundation and contributors
 *
 * SPDX-License-Identifier: MIT
 */

import Foundation
import Capacitor

@objc(SQLiteVanillaPlugin)
public class SQLiteVanillaPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SQLiteVanillaPlugin"
    public let jsName = "SQLiteVanilla"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "close", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "execute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "run", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "query", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "executeSet", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isDBOpen", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteDatabase", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVersion", returnType: CAPPluginReturnPromise)
    ]

    private var databases: [String: Database] = [:]
    private let lock = NSLock()

    private func getDB(_ name: String) -> Database? {
        lock.lock()
        defer { lock.unlock() }
        return databases[name]
    }

    private func setDB(_ name: String, _ db: Database) {
        lock.lock()
        defer { lock.unlock() }
        databases[name] = db
    }

    private func removeDB(_ name: String) -> Database? {
        lock.lock()
        defer { lock.unlock() }
        return databases.removeValue(forKey: name)
    }

    @objc func open(_ call: CAPPluginCall) {
        guard let name = call.getString("database") else {
            call.reject("Missing 'database' parameter")
            return
        }
        do {
            if getDB(name) == nil {
                setDB(name, try Database(name: name))
            }
            call.resolve()
        } catch {
            call.reject("Failed to open database: \(error.localizedDescription)")
        }
    }

    @objc func close(_ call: CAPPluginCall) {
        guard let name = call.getString("database") else {
            call.reject("Missing 'database' parameter")
            return
        }
        removeDB(name)?.close()
        call.resolve()
    }

    @objc func execute(_ call: CAPPluginCall) {
        guard let name = call.getString("database"),
              let statements = call.getString("statements") else {
            call.reject("Missing 'database' or 'statements' parameter")
            return
        }
        guard let db = getDB(name) else {
            call.reject("Database '\(name)' is not open")
            return
        }
        do {
            let changes = try db.execute(statements: statements)
            call.resolve([
                "changes": [
                    "changes": changes
                ]
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func run(_ call: CAPPluginCall) {
        guard let name = call.getString("database"),
              let statement = call.getString("statement") else {
            call.reject("Missing 'database' or 'statement' parameter")
            return
        }
        guard let db = getDB(name) else {
            call.reject("Database '\(name)' is not open")
            return
        }
        do {
            let values = parseValues(call.getArray("values"))
            let result = try db.run(statement: statement, values: values)
            call.resolve([
                "changes": [
                    "changes": result.changes,
                    "lastId": result.lastId
                ]
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func query(_ call: CAPPluginCall) {
        guard let name = call.getString("database"),
              let statement = call.getString("statement") else {
            call.reject("Missing 'database' or 'statement' parameter")
            return
        }
        guard let db = getDB(name) else {
            call.reject("Database '\(name)' is not open")
            return
        }
        do {
            let values = parseValues(call.getArray("values"))
            let rows = try db.query(statement: statement, values: values)
            call.resolve(["values": rows])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func executeSet(_ call: CAPPluginCall) {
        guard let name = call.getString("database"),
              let setArray = call.getArray("set") as? [[String: Any]] else {
            call.reject("Missing 'database' or 'set' parameter")
            return
        }
        guard let db = getDB(name) else {
            call.reject("Database '\(name)' is not open")
            return
        }
        let transaction = call.getBool("transaction") ?? true
        do {
            var items: [(statement: String, values: [Any?]?)] = []
            for item in setArray {
                guard let stmt = item["statement"] as? String else {
                    call.reject("Each item in 'set' must have a 'statement' string")
                    return
                }
                let vals = parseValues(item["values"] as? [Any])
                items.append((statement: stmt, values: vals))
            }
            let changes = try db.executeSet(set: items, transaction: transaction)
            call.resolve([
                "changes": [
                    "changes": changes
                ]
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func isDBOpen(_ call: CAPPluginCall) {
        guard let name = call.getString("database") else {
            call.reject("Missing 'database' parameter")
            return
        }
        let isOpen = getDB(name)?.isOpen ?? false
        call.resolve(["result": isOpen])
    }

    @objc func deleteDatabase(_ call: CAPPluginCall) {
        guard let name = call.getString("database") else {
            call.reject("Missing 'database' parameter")
            return
        }
        do {
            if let db = removeDB(name) {
                try db.delete()
            } else {
                let tempDb = try Database(name: name)
                try tempDb.delete()
            }
            call.resolve()
        } catch {
            call.reject("Failed to delete database: \(error.localizedDescription)")
        }
    }

    @objc func getVersion(_ call: CAPPluginCall) {
        guard let name = call.getString("database") else {
            call.reject("Missing 'database' parameter")
            return
        }
        guard let db = getDB(name) else {
            call.reject("Database '\(name)' is not open")
            return
        }
        do {
            let version = try db.getVersion()
            call.resolve(["version": version])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    // MARK: - Helpers

    private func parseValues(_ values: [Any]?) -> [Any?]? {
        guard let values = values else { return nil }
        return values.map { value in
            if value is NSNull {
                return nil
            }
            return value
        }
    }
}

File: android/src/main/java/org/microbit/sqlite/SQLiteVanillaPlugin.java
/**
 * (c) 2026, Micro:bit Educational Foundation and contributors
 *
 * SPDX-License-Identifier: MIT
 */
package org.microbit.sqlite;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "SQLiteVanilla")
public class SQLiteVanillaPlugin extends Plugin {

    private final Map<String, Database> databases = new ConcurrentHashMap<>();

    @PluginMethod
    public void open(PluginCall call) {
        String name = call.getString("database");
        if (name == null) {
            call.reject("Missing 'database' parameter");
            return;
        }
        try {
            databases.computeIfAbsent(name, (k) -> new Database(getContext(), k));
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to open database: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void close(PluginCall call) {
        String name = call.getString("database");
        if (name == null) {
            call.reject("Missing 'database' parameter");
            return;
        }
        Database db = databases.remove(name);
        if (db != null) {
            db.close();
        }
        call.resolve();
    }

    @PluginMethod
    public void execute(PluginCall call) {
        String name = call.getString("database");
        String statements = call.getString("statements");
        if (name == null || statements == null) {
            call.reject("Missing 'database' or 'statements' parameter");
            return;
        }
        Database db = databases.get(name);
        if (db == null) {
            call.reject("Database '" + name + "' is not open");
            return;
        }
        try {
            long changes = db.execute(statements);
            JSObject changesObj = new JSObject();
            changesObj.put("changes", changes);
            JSObject ret = new JSObject();
            ret.put("changes", changesObj);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void run(PluginCall call) {
        String name = call.getString("database");
        String statement = call.getString("statement");
        if (name == null || statement == null) {
            call.reject("Missing 'database' or 'statement' parameter");
            return;
        }
        Database db = databases.get(name);
        if (db == null) {
            call.reject("Database '" + name + "' is not open");
            return;
        }
        try {
            JSONArray values = call.getArray("values");
            long[] result = db.run(statement, values);
            JSObject changes = new JSObject();
            changes.put("changes", result[0]);
            changes.put("lastId", result[1]);
            JSObject ret = new JSObject();
            ret.put("changes", changes);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void query(PluginCall call) {
        String name = call.getString("database");
        String statement = call.getString("statement");
        if (name == null || statement == null) {
            call.reject("Missing 'database' or 'statement' parameter");
            return;
        }
        Database db = databases.get(name);
        if (db == null) {
            call.reject("Database '" + name + "' is not open");
            return;
        }
        try {
            JSONArray values = call.getArray("values");
            List<JSObject> rows = db.query(statement, values);
            JSObject ret = new JSObject();
            ret.put("values", new JSONArray(rows));
            call.resolve(ret);
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void executeSet(PluginCall call) {
        String name = call.getString("database");
        if (name == null) {
            call.reject("Missing 'database' parameter");
            return;
        }
        Database db = databases.get(name);
        if (db == null) {
            call.reject("Database '" + name + "' is not open");
            return;
        }
        JSONArray setArray = call.getArray("set");
        if (setArray == null) {
            call.reject("Missing 'set' parameter");
            return;
        }
        Boolean transactionParam = call.getBoolean("transaction");
        boolean transaction = transactionParam == null || transactionParam;
        try {
            List<String[]> statements = new ArrayList<>();
            List<JSONArray> valuesList = new ArrayList<>();
            for (int i = 0; i < setArray.length(); i++) {
                JSONObject item = setArray.getJSONObject(i);
                String stmt = item.getString("statement");
                JSONArray vals = item.optJSONArray("values");
                statements.add(new String[] { stmt });
                valuesList.add(vals);
            }
            long totalChanges = db.executeSet(statements, valuesList, transaction);
            JSObject changes = new JSObject();
            changes.put("changes", totalChanges);
            JSObject ret = new JSObject();
            ret.put("changes", changes);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }

    @PluginMethod
    public void isDBOpen(PluginCall call) {
        String name = call.getString("database");
        if (name == null) {
            call.reject("Missing 'database' parameter");
            return;
        }
        Database db = databases.get(name);
        boolean isOpen = db != null && db.isOpen();
        JSObject ret = new JSObject();
        ret.put("result", isOpen);
        call.resolve(ret);
    }

    @PluginMethod
    public void deleteDatabase(PluginCall call) {
        String name = call.getString("database");
        if (name == null) {
            call.reject("Missing 'database' parameter");
            return;
        }
        try {
            Database db = databases.remove(name);
            if (db != null) {
                db.delete(getContext());
            } else {
                Database tempDb = new Database(getContext(), name);
                tempDb.delete(getContext());
            }
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to delete database: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void getVersion(PluginCall call) {
        String name = call.getString("database");
        if (name == null) {
            call.reject("Missing 'database' parameter");
            return;
        }
        Database db = databases.get(name);
        if (db == null) {
            call.reject("Database '" + name + "' is not open");
            return;
        }
        try {
            int version = db.getVersion();
            JSObject ret = new JSObject();
            ret.put("version", version);
            call.resolve(ret);
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }
}
